src/main/scala/it/pagopa/interop/authorizationmanagement/model/persistence/projection/ClientCqrsProjection.scala

```diff
@@ -98,6 +98,38 @@ object ClientCqrsProjection {
           Updates.set("data.purposes.$[elem].agreement.agreementId", agreementId.toString)
         )
       )
+    case AgreementAndEServiceStatesUpdated(
+          eserviceId,
+          descriptorId,
+          consumerId,
+          agreementId,
+          agreementState,
+          eserviceState,
+          audience,
+          voucherLifespan
+        ) =>
+      ActionWithBson(
+        collection.updateMany(
+          Filters.empty(),
+          _,
+          UpdateOptions().arrayFilters(
+            List(
+              Filters.and(
+                Filters.eq("elem.agreement.eserviceId", eserviceId),
+                Filters.eq("elem.agreement.consumerId", consumerId)
+              )
+            ).asJava
+          )
+        ),
+        Updates.combine(
+          Updates.set("data.purposes.$[elem].agreement.state", agreementState.toString),
+          Updates.set("data.purposes.$[elem].agreement.agreementId", agreementId.toString),
+          Updates.set("data.purposes.$[elem].eService.descriptorId", descriptorId.toString),
+          Updates.set("data.purposes.$[elem].eService.audience", audience),
+          Updates.set("data.purposes.$[elem].eService.voucherLifespan", voucherLifespan),
+          Updates.set("data.purposes.$[elem].eService.state", eserviceState.toString)
+        )
+      )
```

Comment:
This is a case where I found little clarity, but it's just this case.

Reply:
It may be worth refactoring, though. You may be right.

Reply:
IMO not for these projections, you'll see that in other services it gets worse instead of better.
While for the behaviour we have a nice improvement 👍